### PR TITLE
feat: Add Event handler

### LINKED ISSUE #44: Event Handling
For each event type, we need to handle them appropriately.

Replace has already been done as a base example in `server/main.py`. An EventHandler class would be best to abstract this away and keep the code in the websocket function clean and tidy. The data should be casted to the correct XData model to not cause typing issues.

Send the client id on connect and sync events for the front end to distinguish between users with the same username.

### PR DESCRIPTION
**Pull request by LukeHankey**
Fixes #44 

Doesn't include the Sync event as not sure we receive that from the frontend. May be best to add that in a different PR as it is not in scope.

### CODE REVIEW

**Review by bensgilbert — APPROVED** [review]
👌

File: server/events.py [hunk 1]
```diff
@@ -146,3 +152,66 @@ class EventResponse(EventRequest):
...
+
+    def __init__(self, client: Client, connection: ConnectionManager):
+        """Initializes the event handler for each client.
+
+        Args:
+            client: The client sending the requests.
+            connection: The connection to the room.
+        """
+        self.client = client
+        self.connection = connection
```
Comment (kappq) [inline server/events.py]: I would rename this as `manager`: a connection basically corresponds to a client.

File: server/connection_manager.py [hunk 2]
```diff
@@ -100,7 +100,7 @@ def update_code_cache(self, room_code: str, replace_data: ReplaceData) -> None:
                 self._rooms[room_code]["code"] = updated_code
 
     async def broadcast(
-        self, data: SyncData | ReplaceData, room_code: str, sender: Client | None = None, buggy: bool = False
+        self, data: type[EventData], room_code: str, sender: Client | None = None, buggy: bool = False
```
Comment (kappq) [inline server/connection_manager.py:103]: `broadcast` still accepts a lot of parameters. As I said in a previous comment, I probably wouldn't leave all this work to `broadcast`. shouldn't this happen in the event handler?

Comment (LukeHankey) [reply server/connection_manager.py:103]: Can you elaborate? What shouldn't be in the broadcast?

Comment (kappq) [reply server/connection_manager.py:103]: > Can you elaborate? What shouldn't be in the broadcast?

In my opinion `broadcast` should only send the message to every client in a room, in this case, it is also modifying the code and updating the code cache. I would leave that to the event handler.

Comment (LukeHankey) [reply server/connection_manager.py:103]: > > Can you elaborate? What shouldn't be in the broadcast?
> 
> In my opinion `broadcast` should only send the message to every client in a room, in this case, it is also modifying the code and updating the code cache. I would leave that to the event handler.

Oh, I thought we were going to put that in an Editor class? That's why I left that logic in there for another PR

Comment (kappq) [reply server/connection_manager.py:103]: Right, the code would still be moved out of `broadcast` though, but I think it's fine if we leave that out for another PR.

**Review by kappq — APPROVED** [review]

**Pull request merged**